File: java/src/main/java/freetranslator/DecisionMatrix.java
package freetranslator;

import java.util.*;
import java.util.stream.*;
import java.util.function.*;

/**
 * FreeTranslator v2.1 – Kettős Réteg Döntési Mátrix (Java port)
 *
 * BELSŐ réteg (fusion/mag):
 *   Minden változó normalizált × súly → összeg → szög → domináns irány
 *
 * KÜLSŐ réteg (decay/héj):
 *   Minden változó affinitással sugározza energiáját a 4 végpont felé
 *
 * Összehasonlítás: rezonancia / divergencia mérték
 */
public class DecisionMatrix {

    public static final List<String> DIRS = List.of("north", "east", "south", "west");
    private static final Map<String, Integer> DIR_ANGLES = Map.of(
        "north", 0, "east", 90, "south", 180, "west", 270
    );

    // ── Változó katalógus ────────────────────────────────────────────────────

    public record Variable(
        String id, String label, String layer, double w0,
        ToDoubleFunction<Double> norm,
        Function<Double, Map<String, Double>> affinity
    ) {}

    private static Map<String, Double> dir4(double n, double e, double s, double w) {
        return Map.of("north", n, "east", e, "south", s, "west", w);
    }

    public static final List<Variable> CATALOG = List.of(
        new Variable("reticleAngle", "Irányzék szög (°)", "both", 0.85,
            v -> v / 360.0,
            v -> { double r = Math.toRadians(v);
                   return dir4(Math.max(0,Math.cos(r)), Math.max(0,Math.cos(r-Math.PI/2)),
                               Math.max(0,Math.cos(r-Math.PI)), Math.max(0,Math.cos(r-3*Math.PI/2))); }),
        new Variable("north", "Észak %", "outer", 0.72,
            v -> v / 100.0,
            v -> dir4(v/100.0, 0.05, 0.05, 0.05)),
        new Variable("east", "Kelet %", "outer", 0.91,
            v -> v / 100.0,
            v -> dir4(0.05, v/100.0, 0.05, 0.05)),
        new Variable("south", "Dél %", "outer", 0.54,
            v -> v / 100.0,
            v -> dir4(0.05, 0.05, v/100.0, 0.05)),
        new Variable("west", "Nyugat %", "outer", 0.31,
            v -> v / 100.0,
            v -> dir4(0.05, 0.05, 0.05, v/100.0)),
        new Variable("matrixEnergy", "Mátrix energia", "inner", 0.78,
            v -> v / 100.0,
            v -> { double en = v/100.0;
                   return dir4(en*0.5, en*0.4, (1-en)*0.4, (1-en)*0.2); }),
        new Variable("tension", "Feszültség", "inner", 0.65,
            v -> v,
            v -> dir4(0.1, 0.1, v*0.7, (1-v)*0.5)),
        new Variable("resonance", "Rezonancia", "inner", 0.45,
            v -> v,
            v -> dir4(v*0.3, v*0.5, v*0.1, v*0.4)),
        new Variable("vibFreqHz", "Rezgés (Hz)", "both", 0.67,
            v -> Math.min(1, Math.log10(Math.max(1, v)) / 5.0),
            v -> { double f = Math.min(1, v/20000.0);
                   return dir4(f*0.6, f*0.3, (1-f)*0.5, (1-f)*0.3); }),
        new Variable("vibAmp", "Rezgés amp", "both", 0.42,
            v -> Math.min(1, v),
            v -> dir4(v*0.4, v*0.4, (1-v)*0.3, (1-v)*0.4)),
        new Variable("contextSize", "Kontextus méret", "both", 0.38,
            v -> Math.min(1, v/20.0),
            v -> { double c = Math.min(1, v/20.0);
                   return dir4(0.1, c*0.6, 0.05, c*0.8); }),
        new Variable("step", "Lépésszám", "both", 0.20,
            v -> Math.min(1, v/1000.0),
            v -> { double s = Math.min(1, v/1000.0);
                   return dir4(0.1, 0.1, s*0.3, s*0.6); })
    );

    // Súly tábla (tanulással módosítható)
    private final Map<String, Double> weights = new HashMap<>();

    public DecisionMatrix() {
        CATALOG.forEach(v -> weights.put(v.id(), v.w0()));
    }

    public double getWeight(String id) { return weights.getOrDefault(id, 0.5); }

    public void adaptWeight(String id, double delta) {
        weights.merge(id, delta, (old, d) -> Math.min(1, Math.max(0, old + d)));
    }

    public Map<String, Double> getAllWeights() { return Collections.unmodifiableMap(weights); }

    // ── Kiértékelt változó ───────────────────────────────────────────────────

    public record EvalVar(
        String id, String label, String layer,
        double raw, double normalized, double weight,
        Map<String, Double> affinity
    ) {}

    private EvalVar eval(Variable v, Map<String, Double> values) {
        double raw  = values.getOrDefault(v.id(), 0.0);
        double nv   = v.norm().applyAsDouble(raw);
        if (!Double.isFinite(nv)) nv = 0;
        return new EvalVar(v.id(), v.label(), v.layer(), raw, nv, getWeight(v.id()), v.affinity().apply(raw));
    }

    // ── BELSŐ RÉTEG – fusion ─────────────────────────────────────────────────

    public record InnerResult(
        double score, double angleOut, String dominant,
        double confidence, List<Map<String, Object>> contributions
    ) {}

    public InnerResult computeInner(List<EvalVar> vars) {
        double weightedSum = 0, totalWeight = 0;
        List<Map<String, Object>> contribs = new ArrayList<>();

        for (EvalVar ev : vars) {
            if ("outer".equals(ev.layer())) continue;
            double contrib = ev.normalized() * ev.weight();
            weightedSum += contrib;
            totalWeight += ev.weight();
            contribs.add(Map.of("id", ev.id(), "contrib", contrib, "weight", ev.weight()));
        }

        double score      = totalWeight > 0 ? weightedSum / totalWeight : 0;
        double angleOut   = score * 360.0;
        String dominant   = findDominant(angleOut);
        double confidence = Math.abs(score - 0.5) * 2.0;

        contribs.sort(Comparator.comparingDouble(m -> -((Double) m.get("contrib"))));
        return new InnerResult(round6(score), round2(angleOut), dominant, round4(confidence), contribs);
    }

    private String findDominant(double angleDeg) {
        double a = MathEngine.normalizeAngle(angleDeg);
        return DIRS.stream().min(Comparator.comparingDouble(d -> {
            double diff = Math.abs(a - DIR_ANGLES.get(d));
            return Math.min(diff, 360 - diff);
        })).orElse("north");
    }

    // ── KÜLSŐ RÉTEG – decay ──────────────────────────────────────────────────

    public record OuterResult(
        Map<String, Integer> distribution, String dominant,
        double dominantScore, Map<String, Double> rawScores
    ) {}

    public OuterResult computeOuter(List<EvalVar> vars) {
        Map<String, Double> scores  = new HashMap<>(Map.of("north",0d,"east",0d,"south",0d,"west",0d));

        for (EvalVar ev : vars) {
            if ("inner".equals(ev.layer())) continue;
            double totalAff = DIRS.stream().mapToDouble(d -> ev.affinity().getOrDefault(d, 0d)).sum();
            if (totalAff == 0) continue;
            for (String dir : DIRS) {
                double aff    = ev.affinity().getOrDefault(dir, 0d) / totalAff;
                double contrib = ev.normalized() * ev.weight() * aff;
                scores.merge(dir, contrib, Double::sum);
            }
        }

        double total = scores.values().stream().mapToDouble(v -> v).sum();
        Map<String, Integer> dist = new LinkedHashMap<>();
        DIRS.forEach(d -> dist.put(d, total > 0 ? (int) Math.round(scores.get(d) / total * 100) : 25));

        // Korrekció: összeg = 100
        String maxDir = DIRS.stream().max(Comparator.comparingInt(dist::get)).orElse("north");
        dist.put(maxDir, dist.get(maxDir) + 100 - dist.values().stream().mapToInt(i -> i).sum());

        String dominant    = DIRS.stream().max(Comparator.comparingInt(dist::get)).orElse("north");
        double domScore    = dist.get(dominant) / 100.0;

        return new OuterResult(dist, dominant, round4(domScore), Map.copyOf(scores));
    }

    // ── ÖSSZEHASONLÍTÁS ──────────────────────────────────────────────────────

    public record Comparison(
        boolean agree, double resonance, double angleDiff,
        double fusionConf, double decayConf, double jointConf,
        boolean anomaly, String verdict,
        String innerDominant, String outerDominant
    ) {}

    public Comparison compare(InnerResult inner, OuterResult outer) {
        boolean agree     = inner.dominant().equals(outer.dominant());
        double iAngle     = DIR_ANGLES.get(inner.dominant());
        double oAngle     = DIR_ANGLES.get(outer.dominant());
        double angleDiff  = Math.abs(iAngle - oAngle);
        if (angleDiff > 180) angleDiff = 360 - angleDiff;

        double resonance  = 1.0 - angleDiff / 180.0;
        double jointConf  = (inner.confidence() + outer.dominantScore()) / 2.0;
        boolean anomaly   = resonance < 0.33;

        String verdict = anomaly
            ? String.format("DIVERGENCIA – belső: %s, külső: %s", inner.dominant(), outer.dominant())
            : String.format("REZONANCIA – %s (%.0f%%)", inner.dominant(), resonance * 100);

        return new Comparison(agree, round4(resonance), angleDiff,
            round4(inner.confidence()), round4(outer.dominantScore()), round4(jointConf),
            anomaly, verdict, inner.dominant(), outer.dominant());
    }

    // ── Fő compute ───────────────────────────────────────────────────────────

    public record DecisionResult(
        long timestamp, InnerResult inner, OuterResult outer,
        Comparison comparison, Map<String, Double> weights
    ) {}

    public DecisionResult compute(Map<String, Double> values) {
        List<EvalVar> vars = CATALOG.stream().map(c -> eval(c, values)).collect(Collectors.toList());
        InnerResult  inner = computeInner(vars);
        OuterResult  outer = computeOuter(vars);
        Comparison   cmp   = compare(inner, outer);

        // Önfejlesztés: rezonáns egyezés erősíti az adaptációs súlyt
        if (cmp.agree() && cmp.resonance() > 0.7) {
            String adaptId = "adapt" + cap(inner.dominant());
            adaptWeight(adaptId, 0.01);
        }
        // Divergencia: felső változók súlyát csökkenti
        if (cmp.anomaly()) {
            inner.contributions().stream().limit(3)
                .forEach(c -> adaptWeight((String) c.get("id"), -0.005));
        }

        return new DecisionResult(System.currentTimeMillis(), inner, outer, cmp, getAllWeights());
    }

    private static String cap(String s) {
        return s.isEmpty() ? s : Character.toUpperCase(s.charAt(0)) + s.substring(1);
    }

    private static double round2(double v) { return Math.round(v * 100.0) / 100.0; }
    private static double round4(double v) { return Math.round(v * 10000.0) / 10000.0; }
    private static double round6(double v) { return Math.round(v * 1000000.0) / 1000000.0; }
}

File: src/decision-matrix.js
/**
 * Kettős réteg döntési mátrix
 *
 * Analógia: atom bomlás (decay) és egyesülés (fusion)
 *
 * BELSŐ réteg – fusion (egyesülés / mag):
 *   Minden változó → normalizált súlyPontszám → egyetlen döntő skalár → domináns irány
 *
 * KÜLSŐ réteg – decay (bomlás / héj):
 *   Minden változó kisugározódik a négy végpont felé → irányvektoros endpoint-eloszlás
 *
 * Összehasonlítás:
 *   A két réteg domináns iránya összevethetó → rezonancia / divergencia mérték
 *
 * @module decision-matrix
 */

'use strict';

// ── Változó katalógus ─────────────────────────────────────────────────────────
// layer: 'inner' | 'outer' | 'both'
// w0:    alap súlyPontszám (0–1); tanulással módosítható
// norm:  [0, max] → [0, 1] normalizáló függvény
// affinity: irányprioritás (outer layer decay súly)
const VARIABLE_CATALOG = [
  {
    id: 'reticleAngle', label: 'Irányzék szög (°)', layer: 'both', w0: 0.85,
    norm: v => v / 360,
    // szögből irány-affinitás: cos-alapú vetítés
    affinity: v => {
      const r = v * Math.PI / 180;
      return {
        north: Math.max(0, Math.cos(r)),
        east:  Math.max(0, Math.cos(r - Math.PI / 2)),
        south: Math.max(0, Math.cos(r - Math.PI)),
        west:  Math.max(0, Math.cos(r - 3 * Math.PI / 2))
      };
    }
  },
  {
    id: 'north', label: 'Észak %', layer: 'outer', w0: 0.72,
    norm: v => v / 100,
    affinity: v => ({ north: v / 100, east: 0.05, south: 0.05, west: 0.05 })
  },
  {
    id: 'east', label: 'Kelet %', layer: 'outer', w0: 0.91,
    norm: v => v / 100,
    affinity: v => ({ north: 0.05, east: v / 100, south: 0.05, west: 0.05 })
  },
  {
    id: 'south', label: 'Dél %', layer: 'outer', w0: 0.54,
    norm: v => v / 100,
    affinity: v => ({ north: 0.05, east: 0.05, south: v / 100, west: 0.05 })
  },
  {
    id: 'west', label: 'Nyugat %', layer: 'outer', w0: 0.31,
    norm: v => v / 100,
    affinity: v => ({ north: 0.05, east: 0.05, south: 0.05, west: v / 100 })
  },
  {
    id: 'matrixEnergy', label: 'Mátrix energia', layer: 'inner', w0: 0.78,
    norm: v => v / 100,
    affinity: v => {
      // Magas energia → Észak (kérdések) és Kelet (fordítás)
      const e = v / 100;
      return { north: e * 0.5, east: e * 0.4, south: (1 - e) * 0.4, west: (1 - e) * 0.2 };
    }
  },
  {
    id: 'tension', label: 'Feszültség', layer: 'inner', w0: 0.65,
    norm: v => v,
    affinity: v => ({ north: 0.1, east: 0.1, south: v * 0.7, west: (1 - v) * 0.5 })
  },
  {
    id: 'resonance', label: 'Rezonancia', layer: 'inner', w0: 0.45,
    norm: v => v,
    affinity: v => ({ north: v * 0.3, east: v * 0.5, south: v * 0.1, west: v * 0.4 })
  },
  {
    id: 'vibFreqHz', label: 'Rezgés (Hz)', layer: 'both', w0: 0.67,
    norm: v => Math.min(1, Math.log10(Math.max(1, v)) / 5),
    affinity: v => {
      const f = Math.min(1, v / 20000);
      return { north: f * 0.6, east: f * 0.3, south: (1 - f) * 0.5, west: (1 - f) * 0.3 };
    }
  },
  {
    id: 'vibAmp', label: 'Rezgés amp', layer: 'both', w0: 0.42,
    norm: v => Math.min(1, v),
    affinity: v => ({ north: v * 0.4, east: v * 0.4, south: (1 - v) * 0.3, west: (1 - v) * 0.4 })
  },
  {
    id: 'balloonRadius', label: 'Burok sugár', layer: 'inner', w0: 0.58,
    norm: v => Math.min(1, v / 0.7),
    affinity: v => {
      const r = Math.min(1, v / 0.7);
      return { north: r * 0.3, east: r * 0.5, south: (1 - r) * 0.4, west: (1 - r) * 0.3 };
    }
  },
  {
    id: 'contextSize', label: 'Kontextus méret', layer: 'both', w0: 0.38,
    norm: v => Math.min(1, v / 20),
    affinity: v => {
      const c = Math.min(1, v / 20);
      return { north: 0.1, east: c * 0.6, south: 0.05, west: c * 0.8 }; // Ny = tanulás
    }
  },
  {
    id: 'adaptNorth', label: 'Adapt É-súly', layer: 'inner', w0: 0.33,
    norm: v => Math.min(1, Math.max(0, v - 1)),
    affinity: v => ({ north: Math.max(0, v - 1), east: 0.05, south: 0.02, west: 0.02 })
  },
  {
    id: 'adaptEast', label: 'Adapt K-súly', layer: 'inner', w0: 0.48,
    norm: v => Math.min(1, Math.max(0, v - 1)),
    affinity: v => ({ north: 0.05, east: Math.max(0, v - 1), south: 0.02, west: 0.02 })
  },
  {
    id: 'adaptSouth', label: 'Adapt D-súly', layer: 'inner', w0: 0.29,
    norm: v => Math.min(1, Math.max(0, v - 1)),
    affinity: v => ({ north: 0.02, east: 0.05, south: Math.max(0, v - 1), west: 0.02 })
  },
  {
    id: 'adaptWest', label: 'Adapt Ny-súly', layer: 'inner', w0: 0.25,
    norm: v => Math.min(1, Math.max(0, v - 1)),
    affinity: v => ({ north: 0.02, east: 0.02, south: 0.02, west: Math.max(0, v - 1) })
  },
  {
    id: 'step', label: 'Lépésszám', layer: 'both', w0: 0.20,
    norm: v => Math.min(1, v / 1000),
    affinity: v => {
      const s = Math.min(1, v / 1000);
      return { north: 0.1, east: 0.1, south: s * 0.3, west: s * 0.6 }; // idő → tanulás
    }
  }
];

const DIRS = ['north', 'east', 'south', 'west'];
const DIR_LABELS = { north: 'Észak (0°)', east: 'Kelet (90°)', south: 'Dél (180°)', west: 'Nyugat (270°)' };

// ── Súlyok tanulással módosítható táblája ─────────────────────────────────────
const _weights = {};
VARIABLE_CATALOG.forEach(v => { _weights[v.id] = v.w0; });

function getWeight(id)          { return _weights[id] ?? 0.5; }
function adaptWeight(id, delta) { _weights[id] = Math.min(1, Math.max(0, (_weights[id] ?? 0.5) + delta)); }
function getAllWeights()         { return { ..._weights }; }

// ── Normalizálás és affinitás kiértékelés ────────────────────────────────────
function evalVariable(catalog, values) {
  const id  = catalog.id;
  const raw = values[id] ?? 0;
  const nv  = isFinite(catalog.norm(raw)) ? catalog.norm(raw) : 0;
  const w   = getWeight(id);
  const aff = catalog.affinity(raw);
  return { id, label: catalog.label, layer: catalog.layer, raw, normalized: +nv.toFixed(4), weight: +w.toFixed(4), affinity: aff };
}

// ── BELSŐ RÉTEG – fusion / egyesülés ─────────────────────────────────────────
/**
 * Minden változó normalizált értékét × súlyával összegzi.
 * Az összeg a [0, 360] tartományba vetítve irányszöget ad.
 * Ez a "mag" döntés – egyetlen pont.
 */
function computeInnerLayer(variables) {
  let weightedSum = 0;
  let totalWeight = 0;

  const contributions = [];

  for (const ev of variables) {
    if (ev.layer === 'outer') continue; // csak 'inner' + 'both'
    const contrib = ev.normalized * ev.weight;
    weightedSum  += contrib;
    totalWeight  += ev.weight;
    contributions.push({ id: ev.id, label: ev.label, contrib: +contrib.toFixed(5), weight: ev.weight, norm: ev.normalized });
  }

  const score       = totalWeight > 0 ? weightedSum / totalWeight : 0;
  const angleOut    = score * 360;
  const dominant    = findDominantAngle(angleOut);
  const confidence  = Math.abs(score - 0.5) * 2; // 0 = semleges, 1 = erős

  contributions.sort((a, b) => b.contrib - a.contrib);

  return { score: +score.toFixed(6), angleOut: +angleOut.toFixed(2), dominant, confidence: +confidence.toFixed(4), contributions };
}

function findDominantAngle(angleDeg) {
  const a = ((angleDeg % 360) + 360) % 360;
  const candidates = [
    { dir: 'north', angle: 0   },
    { dir: 'east',  angle: 90  },
    { dir: 'south', angle: 180 },
    { dir: 'west',  angle: 270 }
  ];
  return candidates.reduce((best, c) => {
    const d = Math.min(Math.abs(a - c.angle), 360 - Math.abs(a - c.angle));
    const bd = Math.min(Math.abs(a - best.angle), 360 - Math.abs(a - best.angle));
    return d < bd ? c : best;
  }).dir;
}

// ── KÜLSŐ RÉTEG – decay / bomlás ─────────────────────────────────────────────
/**
 * Minden változó kisugározza az energiáját a négy végpont felé, affinity-súlyozással.
 * Az összesített végpont-energiák adják a külső réteg eloszlását.
 */
function computeOuterLayer(variables) {
  const endpointScores = { north: 0, east: 0, south: 0, west: 0 };
  const endpointWeights = { north: 0, east: 0, south: 0, west: 0 };
  const varContribs = {};

  for (const ev of variables) {
    if (ev.layer === 'inner') continue; // csak 'outer' + 'both'
    const totalAff = DIRS.reduce((s, d) => s + (ev.affinity[d] || 0), 0);
    if (totalAff === 0) continue;

    const varRow = {};
    for (const dir of DIRS) {
      const aff     = (ev.affinity[dir] || 0) / totalAff;
      const contrib = ev.normalized * ev.weight * aff;
      endpointScores[dir]  += contrib;
      endpointWeights[dir] += ev.weight;
      varRow[dir] = +contrib.toFixed(5);
    }
    varContribs[ev.id] = varRow;
  }

  // Normalizálás 0-100%
  const totalScore = DIRS.reduce((s, d) => s + endpointScores[d], 0);
  const distribution = {};
  for (const dir of DIRS) {
    distribution[dir] = totalScore > 0 ? Math.round(endpointScores[dir] / totalScore * 100) : 25;
  }

  // Korrekció: összeg legyen pontosan 100
  const corrDir = DIRS.reduce((a, b) => distribution[a] > distribution[b] ? a : b);
  distribution[corrDir] += 100 - DIRS.reduce((s, d) => s + distribution[d], 0);

  const dominant = DIRS.reduce((a, b) => distribution[a] > distribution[b] ? a : b);
  const domScore = distribution[dominant] / 100;

  return { distribution, dominant, dominantScore: +domScore.toFixed(4), rawScores: { ...endpointScores }, varContribs };
}

// ── ÖSSZEHASONLÍTÁS – rezonancia mérték ──────────────────────────────────────
/**
 * Összeméri a belső és külső réteg domináns irányát.
 * Rezonancia = 1.0 → teljes egyezés
 * Rezonancia = 0.0 → maximális divergencia (ellentétes irányok)
 */
function compareLayersResult(inner, outer) {
  const agree   = inner.dominant === outer.dominant;

  // Szögtávolság a két domináns irány között
  const ANGLES  = { north: 0, east: 90, south: 180, west: 270 };
  const iAngle  = ANGLES[inner.dominant];
  const oAngle  = ANGLES[outer.dominant];
  let angleDiff = Math.abs(iAngle - oAngle);
  if (angleDiff > 180) angleDiff = 360 - angleDiff;

  const resonance = 1 - angleDiff / 180; // 0 (ellentétes) – 1 (egyező)

  // Döntési biztonság: ha mindkét réteg magas konfidenciával ugyanoda mutat
  const fusionConf  = inner.confidence;
  const decayConf   = outer.dominantScore;
  const jointConf   = (fusionConf + decayConf) / 2;

  // Anomália: ha a két réteg ellentétesek → instabil rendszer
  const anomaly = resonance < 0.33;

  return {
    agree, resonance: +resonance.toFixed(4), angleDiff,
    fusionConf: +fusionConf.toFixed(4), decayConf: +decayConf.toFixed(4),
    jointConf: +jointConf.toFixed(4), anomaly,
    innerDominant: inner.dominant, outerDominant: outer.dominant,
    verdict: anomaly
      ? `DIVERGENCIA – belső: ${DIR_LABELS[inner.dominant]}, külső: ${DIR_LABELS[outer.dominant]}`
      : `REZONANCIA – ${DIR_LABELS[inner.dominant]} (${(resonance * 100).toFixed(0)}%)`
  };
}

// ── Fő compute függvény ───────────────────────────────────────────────────────
/**
 * @param {object} values  – minden változó aktuális értéke (id → szám)
 * @returns {DecisionResult}
 */
function compute(values) {
  const variables = VARIABLE_CATALOG.map(c => evalVariable(c, values));

  const inner      = computeInnerLayer(variables);
  const outer      = computeOuterLayer(variables);
  const comparison = compareLayersResult(inner, outer);

  // Önfejlesztés: ha a két réteg egyezik, erősítsd a domináns irány adaptációs súlyát
  if (comparison.agree && comparison.resonance > 0.7) {
    const adaptId = `adapt${inner.dominant.charAt(0).toUpperCase() + inner.dominant.slice(1)}`;
    adaptWeight(adaptId, 0.01);
  }
  // Ha divergál, csökkentsd a túl erős változók súlyát
  if (comparison.anomaly) {
    inner.contributions.slice(0, 3).forEach(c => adaptWeight(c.id, -0.005));
  }

  return {
    timestamp: Date.now(),
    variables: variables.map(v => ({
      id: v.id, label: v.label, layer: v.layer,
      raw: v.raw, normalized: v.normalized, weight: v.weight,
      weightScore: +(v.normalized * v.weight).toFixed(5)
    })),
    inner,
    outer,
    comparison,
    weights: getAllWeights()
  };
}

module.exports = { compute, adaptWeight, getAllWeights, VARIABLE_CATALOG, DIR_LABELS };
